keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,757 @@
++ _COSI_ - Container _Object_ Store Interface, modeled after CSI.
++ _cosi-node-adapter_ - a pod per node which receives Kubelet gRPC _NodePublishVolume_ and _NodeUnpublishVolume_ requests, ensures the target bucket has been provisioned, and notifies the kubelet when the pod can be run.
++ _driver_ - a container (usually paired with a sidecar container) that is responsible for communicating with the underlying object store to create, delete, grant access to, and revoke access from buckets. Drivers talk gRPC and need no knowledge of Kubernetes. Drivers are typically written by storage vendors, and should not be given any access outside of their namespace.
++ _driverless_ - a use-case where no driver exists to support the backend object store. Some COSI automation may still be provided, but backend operations such as creating a bucket or minting credentials will not occur.
++ _greenfield bucket_ - a new backend bucket created by COSI.
++ _green-to-browfield_ - a use-case where COSI creates a new bucket on behalf of a user, and then supports ways for other users in the cluster to share this bucket.
++ _provisioner_ - a generic term meant to describe the combination of a sidecar and driver. "Provisioning" a bucket can mean creating a new bucket or granting access to an existing bucket.
++ _sidecar_ - a Kubernetes-aware container (usually paired with a driver) that is responsible for fullfilling COSI requests with the driver via gRPC calls. The sidecar's access can be restricted to the namespace where it resides, which is expected to be the same namespace as the provisioner. The COSI sidecar is provided by the Kubernetes community.
++ _static provisioning_ - custom resource creation is done by the admin rather than via COSI automation. This may also include _driverless_ but they are independent.
+
+# Proposal
+
+## APIs
+
+### Storage APIs
+
+#### BucketRequest
+
+A user facing, namespaced custom resource requesting provisioning of a new bucket, or requesting access to an existing bucket. A `BucketRequest` (BR) lives in the app's namespace.  In addition to a `BucketRequest`, a [BucketAccessRequest](#bucketaccessrequest) is necessary in order to grant credentials to access the bucket. BRs are required for both greenfield and brownfield uses.
+
+There is a 1:1 mapping of a `BucketRequest` and the cluster scoped `Bucket` instance, meaning there is a single `Bucket` instance for every BR.
+
+```
+apiVersion: cosi.io/v1alpha1
+kind: BucketRequest
+metadata:
+  name:
+  namespace:
+  labels:
+    cosi.io/provisioner: [1]
+  finalizers:
+    - cosi.io/finalizer [2]
+spec:
+  protocol: [3]
+  bucketPrefix: [4]
+  bucketClassName: [5]
+  bucketInstanceName: [6]
+status:
+  bucketAvailable: [7]
+```
+
+1. `labels`: added by COSI.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
```

Comment:
Can a user provide labels to be read by the provisioner?

Reply:
Yes. The user can add labels to their BRs and BARs. The admin can add labels to the BCs and BACs, But we do not have a plan to allow the admin to tell COSI to add labels to the generated B and BA. That would be nice but probably after MVP.

Reply:
OK, so I was trying to make sure that the provider code gets the BR/BAR labels when processing the grpc request, so that it will be possible for users to ask for custom configurations from the provisioner (which will take it's own decision to allow or reject). Can we make sure those labels are sent over grpc?

Reply:
That's an interesting point. Basically should the user, through their BR, be able to influence characteristics of the new bucket -- in essence overriding or adding to the BC? 
IMO, if we allow this it should be explicit and not done via labels.
Also, at one point on the KEP we allowed a BR to specific a map for bucket properties. The BC also had this map (separate from _parameters_). COSI would have merged the BR's map with the BC's map such that the BC's overwrote the same key in the BR. This idea was rejected.

Reply:
I know it was rejected, this is why I don't suggest to standardize this feature, but I just want to make sure that COSI will enable providers to bypass the limitation until standards will follow.

Reply:
Whether the mechanism is standard or formal we want to maintain a separation of concern between the user and the admin. Passing labels k-v to the driver can potentially allow a user to override a `BucketClass.parameters` values, or add new values that were omitted in the BC. My vote is to _not_ merge labels into the `Bucket.parameters` field (which is passed as bucket_context to the driver).

Reply:
@jeffvance I am not suggesting it as the ultimate configuration, as it won't be portable, but I want COSI to give the driver's code (through grpc) all the BR/BAR information (including custom labels) - so it will be the provider's decision if to lose some portability in favor of supporting something that the COSI API has not yet incorporated into BR/BAR spec, where creating a Class for every request doesn't make sense. 

For example a COSI-driver that only imports existing cloud buckets - the BR will have to specify the external bucket name to import for it to be usable for users without admins creating classes - yes, not portable to another provider, but it can be portable between clouds and be a solid solution for many simple use cases.

Should COSI really be opinionated to prevent this?? Portability is important but also being open to customization,  at least until the COSI spec will cover these custom cases in a portable way :)

Pulling @saad-ali @wlan0 @bswartz for your opinions

Reply:
I'm open to input from others. My understanding of what you're suggesting is for COSI to support passing user-entered values to the driver. My opinion so far is that this could be used to circumvent bucket settings defined in the BucketClass by the admin. I also see value in allowing the user to define certain things and not have to bother the admin to create new BCs which are almost identical just to support some object store features. 

In a much earlier version of this KEP, we proposed a way to do both: give admin the ultimate say in what is passed to the driver, but allow the user some control. We did this by adding 2 fields to the BC:
.  enforcedConfig: [string]string map
.  defaultConfig:    [string]string map

and one field to the BucketRequest:
.   bucketConfig:  [string]string map

where _enforcedConfig_ could not be overwritten or bypassed, and _bucketConfig_ was merged into (overwriting) _defaultConfig_. This basic idea was rejected, in part, due to portability concerns.

Using labels or annotations as an "informal" way to provide the same feature doesn't remove the concern AFAICT.

Reply:
I don't think we should standardize a way to bypass our proper COSI design.
But I think we should allow COSI drivers to decide if they need more information from the user, and not prevent it.

Reply:
Today in the KEP the driver received _bucket_context_, which is `Bucket.parameters`. COSI can control what's in this field but it's currently the BC's parameters. It sounds like you'd like the driver to be passed in an additional field, something like _user_config_ or _user_parameters_ so the driver sees this field as being different from what's defined in the BC. 

Interested in input from others.